### PR TITLE
Make hover and focus not respond when buttons and fields are disabled.

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

Disabled fields and buttons were responding to hover and focus changes, and they shouldn't.

## Tests

I added the following tests:
 - Hover and focus tests for both outlined and filled text fields for when they are enabled and disabled.
 - Hover and focus tests for ink well.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.
- [X] I am willing to check all the boxes.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [X] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/test/material/input_decorator_test.dart [hunk 1]
```diff
@@ -2045,104 +2045,155 @@ void main() {
...
+            disabledBorder: const OutlineInputBorder(borderSide: BorderSide(color: disabledColor)),
+            border: const OutlineInputBorder(borderSide: BorderSide(color: enabledBorderColor)),
+          ),
         ),
-      ),
-    );
+      );
+    }
 
+    // Filled
```
Comment (johnsonmh) [inline packages/flutter/test/material/input_decorator_test.dart]: Nit: add period.

```suggestion
    // Filled.
```


https://dart.dev/guides/language/effective-dart/documentation#do-format-comments-like-sentences

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 2]
```diff
@@ -2029,8 +2027,8 @@ class _InputDecoratorState extends State<InputDecorator> with TickerProviderStat
       fillColor: _getFillColor(themeData),
       focusColor: _getFocusColor(themeData),
       hoverColor: _getHoverColor(themeData),
-      isFocused: isFocused,
-      isHovering: isHovering,
+      isFocused: isFocused && decoration.enabled,
```
Comment (johnsonmh) [inline packages/flutter/lib/src/material/input_decorator.dart]: Is there anywhere that we actually could have `isFocused` be true when `decoration.enabled` is false? If not, maybe we should add this logic to the `isFocused` getter?

```dart
bool get isFocused => widget.isFocused && decoration.enabled;
```

Comment (johnsonmh) [reply packages/flutter/lib/src/material/input_decorator.dart]: Same question for `isHovering`

**Review by johnsonmh — APPROVED** [review]
LGTM Thanks for the thorough tests Greg!

File: examples/flutter_gallery/lib/demo/material/text_form_field_demo.dart [hunk 3]
```diff
@@ -223,6 +223,7 @@ class TextFormFieldDemoState extends State<TextFormFieldDemo> {
                 ),
                 const SizedBox(height: 24.0),
                 TextFormField(
+                  enabled: false,
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply examples/flutter_gallery/lib/demo/material/text_form_field_demo.dart]: Yeah, I thought I reverted that already.

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 4]
```diff
@@ -1905,13 +1903,13 @@ class _InputDecoratorState extends State<InputDecorator> with TickerProviderStat
   }
 
   Color _getFocusColor(ThemeData themeData) {
-    if (decoration.filled != true) // filled == null same as filled == false
+    if (decoration.filled != true || !decoration.enabled) // filled == null same as filled == false
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/material/input_decorator.dart]: Good idea.  Done.

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 5]
```diff
@@ -1905,13 +1903,13 @@ class _InputDecoratorState extends State<InputDecorator> with TickerProviderStat
...
       return Colors.transparent;
     return decoration.focusColor ?? themeData.inputDecorationTheme?.focusColor ?? themeData.focusColor;
   }
 
   Color _getHoverColor(ThemeData themeData) {
-    if (isFocused || decoration.filled != true) // filled == null same as filled == false
+    if (isFocused || decoration.filled != true || !decoration.enabled) // filled == null same as filled == false
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/material/input_decorator.dart]: Done.

File: packages/flutter/lib/src/material/input_decorator.dart (hunk 2 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/input_decorator.dart]: Great idea.  Done.

File: packages/flutter/test/material/input_decorator_test.dart [hunk 6]
```diff
@@ -2045,104 +2045,155 @@ void main() {
...
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/test/material/input_decorator_test.dart]: I'll take your period and raise you a complete sentence! Done.

File: packages/flutter/test/material/input_decorator_test.dart [hunk 7]
```diff
@@ -2045,104 +2045,155 @@ void main() {
...
+    expect(getContainerColor(tester), equals(disabledColor));
+    await tester.pump(const Duration(seconds: 10));
+    expect(getContainerColor(tester), equals(disabledColor));
+
+    await pumpDecorator(hovering: true, enabled: false);
+    expect(getContainerColor(tester), equals(disabledColor));
+    await tester.pump(const Duration(seconds: 10));
+    expect(getContainerColor(tester), equals(disabledColor));
+
+    // Outlined
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/test/material/input_decorator_test.dart]: Done.

File: packages/flutter/test/material/input_decorator_test.dart (hunk 1 above)
Comment (gspencergoog) [reply packages/flutter/test/material/input_decorator_test.dart]: Done.

**Comment by gspencergoog** [comment]
@fkorotkov What's up with all of our Windows bots?  This isn't the only PR that's just DOA.  They go red immediately upon submitting a change.

**Comment by fkorotkov** [comment]
@gspencergoog  Sorry, just noticed the GitHub notification. It should be working now. Sorry for the inconvenience! Don't hesitate to ping support@cirruslabs.org next time for quicker response.

**Pull request merged**